scripts/filter_multimappers.py

```diff
@@ -0,0 +1,177 @@
+    number of indels. Then, computes the minimum number of indels and returns
+    a list with the alignments that have no more than that minimum number of
+    indels. In addition, it updates the tag 'NH' and 'XI' to match the final
+    number of alignments kept and its identifier respectively.
+
+    Args:
+        alignments (list[pysam.AlignedSegment]): alignments with the same query name
+
+    Retrns:
+        best_alignments (list[pysam.AlignedSegment]): alignments with the less indels
+    """
+    alignment_indels = [(aln, count_indels(alignment=aln)) for aln in alignments]
+    min_indels = min(alignment_indels, key=lambda x: x[1])[1]
+    best_alignments = [indelCount[0] for indelCount in alignment_indels if indelCount[1] == min_indels]
+    
+    for i in range(len(best_alignments)):
+        best_alignments[i].set_tag('NH', len(best_alignments))
+        best_alignments[i].set_tag('XI', i)
+
+    return best_alignments
+
+
+def write_output(alignments: List[pysam.AlignedSegment]) -> None:
+    """Write the output to the standard output (stdout).
+
+    Args:
+        alignments (list[pysam.AlignedSegment]): alignments with the same query name
+    """
+    if len(alignments) == 1:
+        sys.stdout.write(alignments[0].to_string() + '\n')
+    else:
+        for alignment in alignments:
+            sys.stdout.write(alignment.to_string() + '\n')
+
+
+def main(sam_file: Path) -> None:
+    """Filter multimappers by indels count.
+
+    Args:
+        sam_file (str): Path to the input SAM file.
+    """
+    with pysam.AlignmentFile(sam_file, "r") as samfile:
+
+        # Extract header from input SAM file and write it to stdout
+        header = samfile.header
+        sys.stdout.write(str(header))
+
+        current_query = None
+        current_alignments: list[pysam.AlignedSegment] = []
+        
+        for alignment in samfile:
+ 
+            # Skip secondary and supplementary alignments
+            if alignment.is_secondary or alignment.is_supplementary:
+                continue
+
+            if current_query is None:
+                current_query = alignment.query_name
+
+            if current_query != alignment.query_name:
+                if len(current_alignments) == 1:
+                    write_output(alignments=current_alignments)
+                else:
+                    best_alignments = find_best_alignments(alignments=current_alignments)
+                    write_output(alignments=best_alignments)
+
+                # Reset variables
+                current_query = alignment.query_name
+                current_alignments = [alignment]
+            else:
+                current_alignments.append(alignment)
+
+        # Last set of alignments
+        if len(current_alignments) == 1:
+            write_output(alignments=current_alignments)
+        else:
+            best_alignments = find_best_alignments(alignments=current_alignments)
+            write_output(alignments=best_alignments)
```

Comment:
This part occurs twice, so best to put it in a function. Makes the `main()` function cleaner as well. As a general rule, you should try to keep a function below a dozen statements or so, if possible (of course there are many good reasons not to, so it's certainly not a strict rule).

Reply:
Won't it be better that instead of creating a new function I just modify the `write_output` one ? Looking at the code closer, the `write_output` function has the same conditions as this repeated part. The the function would look something like this:

```
def write_output(alignments: List[pysam.AlignedSegment]) -> None:
    """Write the output to the standard output (stdout).

    Args:
        alignments: alignments with the same query name
    """
    if len(alignments) == 1:
        sys.stdout.write(alignments[0].to_string() + '\n')
    else:
        best_alignments = find_best_alignments(alignments=alignments)
        for alignment in best_alignments:
            sys.stdout.write(alignment.to_string() + '\n')
```

Then the main function would look like this:

```
def main(sam_file: Path) -> None:
    """Filter multimappers by indels count.

    Args:
        sam_file: Path to the input SAM file.
    """
    with pysam.AlignmentFile(sam_file, "r") as samfile:

        sys.stdout.write(str(samfile.header))

        current_query = None
        current_alignments: list[pysam.AlignedSegment] = []
        
        for alignment in samfile:
 
            if alignment.is_secondary or alignment.is_supplementary:
                continue

            if current_query is None:
                current_query = alignment.query_name

            if current_query != alignment.query_name:
                write_output(alignments=best_alignments)
                current_query = alignment.query_name
                current_alignments = [alignment]
            else:
                current_alignments.append(alignment)

        write_output(alignments=current_alignments)
```

Reply:
Yes, fine